File: BlogAdminClient/src/app/services/comment/comment.service.ts
import {Inject, Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import { Observable } from 'rxjs';
import {IResponse} from "../../responses/IResponse";
import {HttpHeader} from "../../httpHeader";
import {Comment} from "../../models/Comment"

@Injectable({
  providedIn: 'root'
})
export class CommentService {

    private route = "comments/";

    constructor(private _client: HttpClient,
                @Inject('HOST_URL') private _host: string) {
    }

    getCommentById(commentId: number): Observable<IResponse<Comment>> {
        return this._client.post<IResponse<Comment>>(
            this._host + this.route + "comment/get-by-id",
            commentId,
            {
                headers: HttpHeader.getAuthorizationAndContentTypeHttpHeaders()
            });
    }

    getAllComments(): Observable<IResponse<Comment[]>> {
        return this._client.get<IResponse<Comment[]>>(
            this._host +  "admin/comments/get-all-comments",
            {
                headers: HttpHeader.getAuthorizationAndContentTypeHttpHeaders()
            });
    }

    getCommentsByPostId(postId: number): Observable<IResponse<Comment[]>> {
        return this._client.post<IResponse<Comment[]>>(
            this._host + this.route + "get-comments-by-post-id",
            postId,
            {
                headers: HttpHeader.getContentTypeHttpHeaders()
            });
    }

    getCommentsByAccount(): Observable<IResponse<Comment[]>> {
        return this._client.get<IResponse<Comment[]>>(
            this._host + this.route + "get-comments-by-account",
            {
                headers: HttpHeader.getAuthorizationHttpHeader()
            });
    }

    getCommentsByAccountId(accountId: number): Observable<IResponse<Comment[]>> {
        return this._client.post<IResponse<Comment[]>>(
            this._host + this.route + "",
            accountId,
            {
                headers: HttpHeader.getAuthorizationHttpHeader()
            });
    }

    addComment(text: string, postId: number): Observable<IResponse<Comment[]>> {
        let body = {
            postId: postId,
            text: text
        };
        return this._client.post<IResponse<Comment[]>>(
            this._host + this.route + "comment/add",
            body,
            {
                headers: HttpHeader.getAuthorizationAndContentTypeHttpHeaders()
            });
    }

    updateComment(newText: string, commentId: number): Observable<IResponse> {
        let body = {
            commentId: commentId,
            newText: newText
        };
        return this._client.post<IResponse>(
            this._host + this.route + "comment/update",
            body,
            {
                headers: HttpHeader.getAuthorizationAndContentTypeHttpHeaders()
            });
    }

    removeComment(commentId: number): Observable<IResponse> {
        return this._client.post<IResponse>(
            this._host + this.route + "comment/delete",
            commentId,
            {
                headers: HttpHeader.getAuthorizationAndContentTypeHttpHeaders()
            });
    }
}

File: CorporateBlog/src/CorporateBlog.Domain/Services/CommentService.cs
using Microsoft.Extensions.Hosting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CorporateBlog.Domain.Entities;
using CorporateBlog.Domain;
using CorporateBlog.Models.Dto;

namespace CorporateBlog.Domain.Services
{
    public class CommentService
    {
        private readonly IUnitOfWork _unit;

        public CommentService(IUnitOfWork unit)
        {
            _unit = unit ?? throw new ArgumentNullException(nameof(unit));
        }

        public async Task<CommentDto> GetCommentById(int commentId)
        {
            var comment = await _unit.CommentRepository.GetById(commentId);

            return new()
            {
                Id = comment.Id,
                Text = comment.Text,
                AccountId = comment.AccountId,
                AccountNickname = comment.Account!.Login,
                PostId = comment.PostId,
                Created = comment.Created,
                IsDeleted = comment.IsDeleted
            };
        }

        public async Task<IReadOnlyList<CommentDto>> AddComment(int accountId, int postId, string text)
        {
            var comment = new Comment()
            {
                AccountId = accountId,
                PostId = postId,
                Text = text,
                Created = DateTime.Now
            };

            await _unit.CommentRepository.Add(comment);
            await _unit.SaveChangesAsync();

            return await GetCommentsByPostId(postId);
        }

        public async Task UpdateComment(int commentId, string newText)
        {
            var comment = await _unit.CommentRepository.GetById(commentId);

            comment.Text = newText;
            comment.Created = DateTime.Now;

            await _unit.CommentRepository.Update(comment);
            await _unit.SaveChangesAsync();

        }

        public async Task RemoveComment(int commentId)
        {
            var comment = await _unit.CommentRepository.GetById(commentId);
            comment.IsDeleted = true;

            await _unit.CommentRepository.Update(comment);
            await _unit.SaveChangesAsync();
        }


        public async Task<IReadOnlyList<CommentDto>> GetCommentsByPostId(int postId)
        {
            var comments = await _unit.CommentRepository.GetCommentsByPostId(postId);

            var commentsDto = new List<CommentDto>();

            foreach (var comment in comments)
            {
                commentsDto.Add(new()
                {
                    Id = comment.Id,
                    Text = comment.Text,
                    AccountId = comment.AccountId,
                    AccountNickname = comment.Account!.Login,
                    PostId = comment.PostId,
                    Created = comment.Created
                });
            }
            return commentsDto;
        }



        public async Task<IReadOnlyList<CommentDto>> GetCommentsOnPublishedPostsByAccountId(int accountId)
        {
            var comments = await _unit.CommentRepository.GetCommentsByAccountId(accountId);

            comments = comments.Where(c => c.Post!.Status == PostStatus.Published).ToList();

            var commentsDto = new List<CommentDto>();

            foreach (var comment in comments)
            {
                commentsDto.Add(new()
                {
                    Id = comment.Id,
                    Text = comment.Text,
                    AccountId = comment.AccountId,
                    AccountNickname = comment.Account!.Login,
                    PostId = comment.PostId,
                    PostTitle = comment.Post!.Title,
                    Created = comment.Created
                });
            }
            return commentsDto;
        }

        public async Task<IReadOnlyList<CommentDto>> GetAllComments()
        {
            var comments = await _unit.CommentRepository.GetAll();

            var commentsDto = new List<CommentDto>();

            foreach (var comment in comments)
            {
                commentsDto.Add(new()
                {
                    Id = comment.Id,
                    Text = comment.Text,
                    AccountId = comment.AccountId,
                    AccountNickname = comment.Account!.Login,
                    PostId = comment.PostId,
                    PostTitle = comment.Post!.Title,
                    Created = comment.Created,
                    IsDeleted = comment.IsDeleted
                });
            }
            return commentsDto;
        }

    }
}
